### DISCUSSION TITLE
Bug: 'Depends' object has no attribute 'role'

### DISCUSSION THREAD

**Discussion by brantje** [open-webui/open-webui#6487] [Issues]
# Bug Report

## Important Notes

- **Before submitting a bug report**: Please check the Issues or Discussions section to see if a similar issue or feature request has already been posted. It's likely we're already tracking it! If you’re unsure, start a discussion post first. This will help us efficiently focus on improving the project.

- **Collaborate respectfully**: We value a constructive attitude, so please be mindful of your communication. If negativity is part of your approach, our capacity to engage may be limited. We’re here to help if you’re open to learning and communicating positively. Remember, Open WebUI is a volunteer-driven project managed by a single maintainer and supported by contributors who also have full-time jobs. We appreciate your time and ask that you respect ours.

- **Contributing**: If you encounter an issue, we highly encourage you to submit a pull request or fork the project. We actively work to prevent contributor burnout to maintain the quality and continuity of Open WebUI.

- **Bug reproducibility**: If a bug cannot be reproduced with a `:main` or `:dev` Docker setup, or a pip install with Python 3.11, it may require additional help from the community. In such cases, we will move it to the "issues" Discussions section due to our limited resources. We encourage the community to assist with these issues. Remember, it’s not that the issue doesn’t exist; we need your help!

Note: Please remove the notes above when submitting your post. Thank you for your understanding and support!

----

## Installation Method
Docker

## Environment

- **Open WebUI Version:** 0.3.35
- **Ollama (if applicable):** [e.g., v0.2.0, v0.1.32-rc1]

- **Operating System:** Ubuntu 24.04
- **Browser (if applicable):** Firefox

**Confirmation:**

- [x] I have read and followed all the instructions provided in the README.md.
- [x] I am on the latest version of both Open WebUI and Ollama.
- [x] I have included the browser console logs.
- [x] I have included the Docker container logs.
- [x] I have provided the exact steps to reproduce the bug in the "Steps to Reproduce" section below.

## Expected Behavior:
Script should run fine

## Actual Behavior:
After updating to 0.3.35 (from 0.3.33), the script crashes with the following error:
```
Uh-oh! There was an issue connecting to mcts mistral-nemo:12b.
'Depends' object has no attribute 'role'
```

## Description
After upgrading from 0.3.33 to 0.3.35, the script crashes with an error it can't find the attribute role.

**Bug Summary:**
[Provide a brief but clear summary of the bug]

## Reproduction Details

**Steps to Reproduce:**
Download the function: https://github.com/av/harbor/blob/main/open-webui/extras/mcts.py
Open a chat with a mcts model.
Type something

## Logs and Screenshots

**Browser Console Logs:**
```
Object { error: {…} }
[index.ts:63:11](https://zzz.zzzz.sh/src/lib/apis/streaming/index.ts)
Object { detail: "'Depends' object has no attribute 'role'" }
[Chat.svelte:1756:10](https://zzz.zzzz.sh/src/lib/components/chat/Chat.svelte)
TypeError: e.choices is undefined
    T index.ts:245
[Chat.svelte:1904:13](https://zzz.zz.sh/src/lib/components/chat/Chat.svelte)
Failed to parse response:  TypeError: e.choices is undefined
    S index.ts:287
[index.ts:315:10](https://zz.zzz.sh/src/lib/apis/index.ts)
Array []
[Chat.svelte:1951:11](https://zz.zz.sh/src/lib/components/chat/Chat.svelte)
```
​
**Docker Container Logs:**
``` 
INFO:     172.70.47.114:0 - "GET /api/v1/chats/?page=1 HTTP/1.1" 200 OK
generate_title
visual_tree_of_thoughts.mcts-mistral-nemo:12b
visual_tree_of_thoughts
visual_tree_of_thoughts
Error: 'Depends' object has no attribute 'role'
INFO:     172.70.47.114:0 - "POST /api/task/title/completions HTTP/1.1" 200 OK
INFO:     172.70.47.114:0 - "POST /api/v1/chats/9bcaedb4-868d-4580-8d81-45aea9e18a0c HTTP/1.1" 200 OK
INFO:     172.70.47.114:0 - "GET /api/v1/chats/?page=1 HTTP/1.1" 200 OK
INFO:     172.70.47.114:0 - "GET /api/v1/chats/9bcaedb4-868d-4580-8d81-45aea9e18a0c/tags HTTP/1.1" 200 OK
2024-10-27 19:22:51,021 - function_visual_tree_of_thoughts - DEBUG - Pipe mcts received: {'model': 'visual_tree_of_thoughts.mcts-mistral-nemo:12b', 'messages': [{'role': 'user', 'content': '### Task:\nGenerate 1-3 broad tags categorizing the main themes of the chat history, along with 1-3 more specific subtopic tags.\n\n### Guidelines:\n- Start with high-level domains (e.g. Science, Technology, Philosophy, Arts, Politics, Business, Health, Sports, Entertainment, Education)\n- Consider including relevant subfields/subdomains if they are strongly represented throughout the conversation\n- If content is too short (less than 3 messages) or too diverse, use only ["General"]\n- Use the chat\'s primary language; default to English if multilingual\n- Prioritize accuracy over specificity\n\n### Output:\nJSON format: { "tags": ["tag1", "tag2", "tag3"] }\n\n### Chat History:\n<chat_history>\nUSER: Three playing cards in a row. Can you name them with these clues? There is a two to the right of a king. A diamond will be found to the left of a spade. An ace is to the left of a heart. A heart is to the left of a spade. Now, identify all three cards.\nASSISTANT: \n</chat_history>'}], 'stream': False}
2024-10-27 19:22:51,021 - function_visual_tree_of_thoughts - DEBUG - Question: ### Task:
Generate 1-3 broad tags categorizing the main themes of the chat history, along with 1-3 more specific subtopic 7ags.

### Guidelines:
- Start with high-level domains (e.g. Science, Technology, Philosophy, Arts, Politics, Business, Health, Sports, Entertainment, Education)
- Consider including relevant subfields/subdomains if they are strongly represented throughout the conversation
- If content is too short (less than 3 messages) or too diverse, use only ["General"]
- Use the chat's primary language; default to English if multilingual
- Prioritize accuracy over specificity

### Output:
JSON format: { "tags": ["tag1", "tag2", "tag3"] }

### Chat History:
<chat_history>
USER: Three playing cards in a row. Can you name them with these clues? There is a two to the right of a king. A diamond will be found to the left of a spade. An ace is to the left of a heart. A heart is to the left of a spade. Now, identify all three cards.
ASSISTANT:
</chat_history>
2024-10-27 19:22:51,021 - function_visual_tree_of_thoughts - DEBUG - Progress: Preparing initial thoughts...
generate_chat_tags
visual_tree_of_thoughts.mcts-mistral-nemo:12b
content ### Task:
Generate 1-3 broad tags categorizing the main themes of the chat history, along with 1-3 more specific subtopic tags.

### Guidelines:
- Start with high-level domains (e.g. Science, Technology, Philosophy, Arts, Politics, Business, Health, Sports, Entertainment, Education)
- Consider including relevant subfields/subdomains if they are strongly represented throughout the conversation
- If content is too short (less than 3 messages) or too diverse, use only ["General"]
- Use the chat's primary language; default to English if multilingual
- Prioritize accuracy over specificity

### Output:
JSON format: { "tags": ["tag1", "tag2", "tag3"] }

### Chat History:
<chat_history>
USER: Three playing cards in a row. Can you name them with these clues? There is a two to the right of a king. A diamond will be found to the left of a spade. An ace is to the left of a heart. A heart is to the left of a spade. Now, identify all three cards.
ASSISTANT:
</chat_history>
visual_tree_of_thoughts
visual_tree_of_thoughts
Error: 'NoneType' object is not callable
INFO:     172.70.47.114:0 - "POST /api/task/tags/completions HTTP/1.1" 200 OK
```

## Additional Information

Did something change in the API's? I'll be glad to fix this script and send a PR to harbor.

**Comment by tjbck** [comment]
Unsure what would've caused this issue, continue your troubleshooting journey and keep us updated!

**Comment by adamoutler** [comment]
I had the same issue on 0.3.31.  I updated to 0.3.35 and was able to try it with gemma2:2b and gemma2:9b last night.  Nothing changed since last night when I "Test"ed. When I awoke this morning I tried again and it was not working. 
![Screenshot_20241104-162936](https://github.com/user-attachments/assets/11905833-c43e-4516-b62e-4663ace5ca6d)
![Screenshot_20241104-163007](https://github.com/user-attachments/assets/cee3f62d-4b64-4e2c-b11a-0fb63d8f615f)

**Reply by tjbck** [reply]
You might want to create an issue post on their github repo, keep us updated!

**Comment by adamoutler** [comment]
I agree.  But just to be sure, I traced it to this async ollama call and added some debugging 
![Screenshot_20241104-192054](https://github.com/user-attachments/assets/4ce3698c-efec-4d70-bbda-fcdff5e59598)

Any chance the problem is in OpenWebUI based on this?
```

2024-11-05 00:22:22,651 - function_visual_tree_of_thoughts - DEBUG - Pipe mcts received: {'stream': True, 'model': 'visual_tree_of_thoughts.mcts-gemma2:2b', 'messages': [{'role': 'system', 'content': '\n\nUser Context:\n1. [2024-10-18]. always have BASH available, but Java and Python work as well. To a lesser extent other languages.\nI appreciate tables, decision trees, pictures, and ascii art where applicable.\n'}, {'role': 'user', 'content': 'Kitty litter'}]}
2024-11-05 00:22:22,651 - function_visual_tree_of_thoughts - DEBUG - Question: kitty litter
2024-11-05 00:22:22,651 - function_visual_tree_of_thoughts - DEBUG - Progress: Preparing initial thoughts...
visual_tree_of_thoughts
visual_tree_of_thoughts
Model:
gemma2:2b
messages:
[{'role': 'user', 'content': '\n<instruction>\nAnswer the question below. Do not pay attention to, unexpected casing, punctuation or accent marks.\n</instruction>\n\n<question>\nkitty litter\n</question>\n'}]
Error: 'Depends' object has no attribute 'role'
```

Ollama is defined as
```
from open_webui.apps.ollama import main as ollama
```

**Comment by nschulzke** [comment]
I ran into the same thing. Not an issue with ollama, the script needed to be updated to pass in the user object.

As a quick patch I made a dataclass that has the right shape:

```diff
diff --git a/functions/tree_of_thoughts.py b/functions/tree_of_thoughts_2.py
index af5bac0..042f759 100644
--- a/functions/mcts_original.py
+++ b/functions/mcts_fixed.py
@@ -12,6 +12,7 @@ import math
 import asyncio
 import json
 import re
+import traceback
 
 from typing import (
     List,
@@ -24,6 +25,7 @@ from typing import (
 )
 from open_webui.constants import TASKS
 from open_webui.apps.ollama import main as ollama
+from dataclasses import dataclass
 
 # ==============================================================================
 
@@ -137,6 +139,14 @@ mods = [
 ]
 
 
+@dataclass
+class User:
+    id: str
+    email: str
+    name: str
+    role: str
+
+
 def modify_text(text, percentage):
     if not text:
         return "", {}  # Return empty string and empty mapping if input is empty
@@ -412,6 +422,8 @@ class Pipe:
         self.__model__ = model
         self.__question__ = base_question
         self.__current_event_emitter__ = __event_emitter__
+        logger.debug(f"User: {__user__}")
+        self.__user__ = User(**__user__)
 
         best_answer = None
         best_score = -float("inf")
@@ -486,7 +498,8 @@ class Pipe:
         messages,
     ) -> AsyncGenerator[str, None]:
         response = await ollama.generate_openai_chat_completion(
-            {"model": model, "messages": messages, "stream": True}
+            {"model": model, "messages": messages, "stream": True},
+            user=self.__user__,
         )
 
         async for chunk in response.body_iterator:
@@ -501,7 +514,8 @@ class Pipe:
 
     async def get_completion(self, model: str, messages):
         response = await ollama.generate_openai_chat_completion(
-            {"model": model, "messages": messages, "stream": False}
+            {"model": model, "messages": messages, "stream": False},
+            user=self.__user__,
         )
 
         return self.get_response_content(response)
```

It looks like in a near-future version of Open WebUI we could just bypass the user check in `generate_openai_chat_completion`, but that's not an option in the latest release.

**Reply by graelo** [reply]
Thanks @nschulzke 

I applied your update, just in case, but do you confirm this function still works fine on your side with 0.3.35? On my side, when the function runs, I invariably get `Unclosed client session` errors, and ultimately the function stops with a `Load failed` error.

I won't provide more details at the moment as I don't want to derail the discussion. I just want to evaluate if this occurs only for me. Thank you

**Reply by nschulzke** [reply]
Yes, 0.0.5 of the mcts script plus my patch above works on 0.3.35 of Open WebUI.

**Reply by adamoutler** [reply]
The problem seems to occur when the user personal info is enabled. This addon was originally designed for `(message)`, but it receives `(message, user)`.  That's enough to fix the critical problem.  

I think this issue can be closed unless its possible to detect non-user functions and pass in messages alone for backward compatibility

**Reply by hqnicolas** [reply]
> I ran into the same thing. Not an issue with ollama, the script needed to be updated to pass in the user object.
> 
> As a quick patch I made a dataclass that has the right shape:

Works

**Comment by NetRider88** [comment]
with the latest update there was a compatibility issue 
here is the code that works : 
`"""
title: mcts
author: av
author_url: https://github.com/av
description: mcts - Monte Carlo Tree Search
version: 0.0.5.1.zzxc.v05

Prominant Notice of Change: Modified by Adam Outler to
fix crashing issue when user data is supplied. Fix based
on patch supplied by nschulzke.
https://github.com/open-webui/open-webui/discussions/6487

Added a fix for setting model using Valves

Updated for Open WebUI 0.5 compatibility:
- Updated import paths from open_webui.apps to open_webui.routers
- Added Request parameter for new function signature
- Fixed deprecated function calls
"""

import logging
import random
import math
import asyncio
import json
import re

from typing import (
    List,
    Optional,
    AsyncGenerator,
    Callable,
    Awaitable,
    Generator,
    Iterator,
)
from open_webui.constants import TASKS
from open_webui.utils.chat import generate_chat_completion  # Updated for v0.5
from fastapi import Request  # Required for v0.5
import traceback
from pydantic import BaseModel, Field
from dataclasses import dataclass

# ==============================================================================

name = "mcts"
default_max_children = 2
default_exploration_weight = 1.414
default_max_iterations = 2
default_max_simulations = 2
default_thoughts = 2


@dataclass
class User:
    id: str
    email: str
    name: str
    role: str


# ==============================================================================

thoughts_prompt = """
<instruction>
Give a suggestion on how this answer can be improved.
WRITE ONLY AN IMPROVEMENT SUGGESTION AND NOTHING ELSE.
YOUR REPLY SHOULD BE A SINGLE SENTENCE.
</instruction>

<question>
{question}
</question>

<draft>
{answer}
</draft>
""".strip()

eval_answer_prompt = """
Given the following text:
"{answer}"

How well does it answers this question:
"{question}"

Rate the answer from 1 to 10, where 1 is completely wrong or irrelevant and 10 is a perfect answer.
Reply with a single number between 1 and 10 only. Do not write anything else, it will be discarded.
THINK CAREFULLY AND USE BEST PRACTICES.
""".strip()

analyze_prompt = """
Iteration Analysis:

Original question: {question}
Best answer found: {best_answer}
Best score achieved: {best_score}

Analyze this iteration of the thought process. Consider the following:
1. What aspects of the best answer made it successful?
2. What patterns or approaches led to higher-scoring thoughts?
3. Were there any common pitfalls or irrelevant tangents in lower-scoring thoughts?
4. How can the thought generation process be improved for the next iteration?

Provide a concise analysis and suggest one specific improvement strategy for the next iteration.
""".strip()

update_prompt = """
<instruction>
Your task is to read the question and the answer below, then analyse the given critique.
When you are done - think about how the answer can be improved based on the critique.
WRITE A REVISED ANSWER THAT ADDRESSES THE CRITIQUE. DO NOT WRITE ANYTHING ELSE.
</instruction>
<question>
{question}
</question>
<draft>
{answer}
</draft>
<critique>
{improvements}
</critique>
""".strip()

initial_prompt = """
<instruction>
Answer the question below. Do not pay attention to, unexpected casing, punctuation or accent marks.
</instruction>

<question>
{question}
</question>
"""

# ==============================================================================


def setup_logger():
    logger = logging.getLogger(__name__)
    if not logger.handlers:
        logger.setLevel(logging.DEBUG)
        handler = logging.StreamHandler()
        handler.set_name(name)
        formatter = logging.Formatter(
            "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
        )
        handler.setFormatter(formatter)
        logger.addHandler(handler)
        logger.propagate = False
    return logger


logger = setup_logger()

# ==============================================================================

mods = [
    "capitalize",
    "diacritic",
    "leetspeak",
    "remove_vowel",
]


def modify_text(text, percentage):
    if not text:
        return "", {}  # Return empty string and empty mapping if input is empty

    if not 0 <= percentage <= 100:
        raise ValueError("Percentage must be between 0 and 100")

    words = text.split()
    chars = list(text)
    num_chars_to_modify = max(1, int(len(chars) * (percentage / 100)))
    indices_to_modify = random.sample(range(len(chars)), num_chars_to_modify)
    word_mapping = {}

    for idx in indices_to_modify:
        modification = random.choice(mods)

        # Find the word that contains the current character
        current_length = 0
        for word_idx, word in enumerate(words):
            if current_length <= idx < current_length + len(word):
                original_word = word
                word_start_idx = current_length
                break
            current_length += len(word) + 1  # +1 for the space
        else:
            # If we're here, we're likely dealing with a space or the last character
            continue

        if modification == "capitalize":
            chars[idx] = chars[idx].swapcase()
        elif modification == "diacritic":
            if chars[idx].isalpha():
                diacritics = ["̀", "́", "̂", "̃", "̈", "̄", "̆", "̇", "̊", "̋"]
                chars[idx] = chars[idx] + random.choice(diacritics)
        elif modification == "leetspeak":
            leetspeak_map = {
                "a": "4",
                "e": "3",
                "i": "1",
                "o": "0",
                "s": "5",
                "t": "7",
                "b": "8",
                "g": "9",
                "l": "1",
            }
            chars[idx] = leetspeak_map.get(chars[idx].lower(), chars[idx])
        elif modification == "remove_vowel":
            if chars[idx].lower() in "aeiou":
                chars[idx] = ""

        modified_word = "".join(
            chars[word_start_idx : word_start_idx + len(original_word)]
        )

        if modified_word != original_word:
            # Clean up both the modified word and the original word
            cleaned_modified_word = modified_word.rstrip(".,!?")
            cleaned_original_word = original_word.rstrip(".,!?")
            word_mapping[cleaned_modified_word] = cleaned_original_word

    modified_text = "".join(chars)
    return modified_text, word_mapping


def replace_with_mapping(text, mapping):
    for key, value in mapping.items():
        text = text.replace(key, value)
    return text


# ==============================================================================


def escape_mermaid(text):
    return text.replace('"', "&quot;").replace("(", "&#40;").replace(")", "&#41;")


class Node:
    id: str
    content: str
    parent: Optional["Node"]
    max_children: int
    children: List["Node"]
    visits: int
    value: float

    def __init__(self, **kwargs):
        self.id = "".join(random.choices("abcdefghijklmnopqrstuvwxyz", k=4))
        self.content = kwargs.get("content")
        self.parent = kwargs.get("parent")
        self.exploration_weight = kwargs.get(
            "exploration_weight", default_exploration_weight
        )
        self.max_children = kwargs.get("max_children", default_max_children)
        self.children = []
        self.visits = 0
        self.value = 0

    def add_child(self, child: "Node"):
        child.parent = self
        self.children.append(child)
        return child

    def fully_expanded(self):
        return len(self.children) >= self.max_children

    def uct_value(self):
        epsilon = 1e-6

        return self.value / (
            self.visits + epsilon
        ) + self.exploration_weight * math.sqrt(
            math.log(self.parent.visits) / (self.visits + epsilon)
        )

    def mermaid(self, offset=0, selected=None):
        padding = " " * offset
        msg = f"{padding}{self.id}({self.id}:{self.visits} - {escape_mermaid(self.content[:25])})\n"

        if selected == self.id:
            msg += f"{padding}style {self.id} stroke:#0ff\n"

        for child in self.children:
            msg += child.mermaid(offset + 4, selected)
            msg += f"{padding}{self.id} --> {child.id}\n"

        return msg

    def best_child(self):
        if not self.children:
            return self

        return max(self.children, key=lambda child: child.visits).best_child()


class MCTS:
    question: str
    root: Node
    llm: "Pipe"
    selected: Optional[Node]
    exploration_weight: float

    def __init__(self, **kwargs):
        self.question = kwargs.get("question")
        self.root = kwargs.get("root")
        self.llm = kwargs.get("llm")
        self.selected = None
        self.exploration_weight = kwargs.get(
            "exploration_weight", default_exploration_weight
        )

    async def select(self):
        logger.debug("Selecting node...")
        node = self.root
        while node.children:
            node = self.uct_select(node)
        return node

    async def expand(self, node):
        logger.debug(f"Expanding node {node.id}...")
        await self.llm.progress(f"Thinking about {node.id}...")

        for _ in range(random.randint(default_thoughts, default_thoughts + 1)):
            await self.llm.emit_replace(self.mermaid(node))
            await self.llm.emit_message(f"Thought: ")
            thought = await self.llm.generate_thought(node.content)
            await self.llm.emit_message(f"\n\n---\n\nSolution:\n")

            new_content = await self.llm.update_approach(node.content, thought)
            child = Node(content=new_content, parent=node)
            node.add_child(child)

        return random.choice(node.children)

    async def simulate(self, node):
        logger.debug(f"Simulating node {node.id}...")
        await self.llm.progress(f"Thinking about {node.id}...")
        await self.llm.emit_replace(self.mermaid())

        return await self.llm.evaluate_answer(node.content)

    def backpropagate(self, node, score):
        logger.debug(f"Backpropagating from {node.id}...")
        while node:
            node.visits += 1
            node.value += score
            node = node.parent

    def uct_select(self, node):
        logger.debug(f"Selecting uct {node.id}...")
        return max(node.children, key=lambda child: child.uct_value())

    def best_child(self):
        return self.root.best_child()

    async def search(self, num_simulations):
        logger.debug("Starting search...")

        for _ in range(num_simulations):
            leaf = await self.select()
            self.selected = leaf
            if not leaf.fully_expanded():
                leaf = await self.expand(leaf)
            score = await self.simulate(leaf)
            self.backpropagate(leaf, score)

        return self.selected

    def mermaid(self, selected=None):
        return f"""
```mermaid
graph LR
{self.root.mermaid(0, selected.id if selected else self.selected.id)}
```
"""


# ==============================================================================

EventEmitter = Callable[[dict], Awaitable[None]]


class Pipe:
    __current_event_emitter__: EventEmitter
    __current_node__: Node
    __question__: str
    __model__: str
    __request__: Request

    class Valves(BaseModel):
        MODEL: str = Field(
            default="llama3.2:latest", description="Model to use (model id from ollama)"
        )

    def __init__(self):
        self.valves = self.Valves()

    def pipes(self) -> list[dict[str, str]]:
        out = [
            {"id": f"{name}-{self.valves.MODEL}", "name": f"{name} {self.valves.MODEL}"}
        ]
        logger.debug(f"Available models: {out}")

        return out

    def resolve_model(self, body: dict) -> str:
        model_id = body.get("model")
        without_pipe = ".".join(model_id.split(".")[1:])
        return without_pipe.replace(f"{name}-", "")

    def resolve_question(self, body: dict) -> str:
        return body.get("messages")[-1].get("content").strip()

    async def pipe(
        self,
        body: dict,
        __user__: dict,
        __event_emitter__=None,
        __task__=None,
        __model__=None,
        __request__: Request = None,  # Added for v0.5 compatibility
    ) -> str | Generator | Iterator:
        model = self.resolve_model(body)
        base_question = self.resolve_question(body)

        if __task__ == TASKS.TITLE_GENERATION:
            content = await self.get_completion(model, body.get("messages"))
            return f"{name}: {content}"

        logger.debug(f"Pipe {name} received: {body}")
        question, mapping = modify_text(base_question, 0)
        logger.debug(f"Question: {question}")

        # TODO: concurrency
        self.__model__ = model
        self.__question__ = base_question
        self.__current_event_emitter__ = __event_emitter__
        self.__request__ = __request__  # Store request for v0.5
        logger.debug(f"User: {__user__}")
        self.__user__ = User(**__user__)

        best_answer = None
        best_score = -float("inf")

        await self.progress("Preparing initial thoughts...")

        initial_reply = await self.stream_prompt_completion(
            initial_prompt, question=question
        )

        root = Node(content=initial_reply)
        mcts = MCTS(root=root, llm=self)

        logger.debug("Starting MCTS...")
        for i in range(default_max_iterations):
            logger.debug(f"Iteration {i + 1}/{default_max_iterations}...")

            await mcts.search(default_max_simulations)
            logger.debug(mcts.mermaid())

            best_child = mcts.best_child()
            score = await self.evaluate_answer(best_child.content)

            if score > best_score:
                best_score = score
                best_answer = best_child.content

        await self.emit_replace(mcts.mermaid(best_child))
        await self.emit_message(f"{best_answer}")
        await asyncio.sleep(0.2)
        await self.done()

        return ""

    async def progress(
        self,
        message: str,
    ):
        logger.debug(f"Progress: {message}")
        await self.emit_status("info", message, False)

    async def done(
        self,
    ):
        await self.emit_status("info", "Fin.", True)

    async def emit_message(self, message: str):
        await self.__current_event_emitter__(
            {"type": "message", "data": {"content": message}}
        )

    async def emit_replace(self, message: str):
        await self.__current_event_emitter__(
            {"type": "replace", "data": {"content": message}}
        )

    async def emit_status(self, level: str, message: str, done: bool):
        await self.__current_event_emitter__(
            {
                "type": "status",
                "data": {
                    "status": "complete" if done else "in_progress",
                    "level": level,
                    "description": message,
                    "done": done,
                },
            }
        )

    async def get_streaming_completion(
        self,
        model: str,
        messages,
    ) -> AsyncGenerator[str, None]:
        # Updated for v0.5 - using the new function signature
        response = await generate_chat_completion(
            self.__request__,
            {"model": model, "messages": messages, "stream": True}, 
            self.__user__
        )
        
        # Handle the response based on the new format
        if hasattr(response, 'body_iterator'):
            async for chunk in response.body_iterator:
                for part in self.get_chunk_content(chunk):
                    yield part
        else:
            # Fallback for different response types
            async for chunk in response:
                for part in self.get_chunk_content(chunk):
                    yield part

    async def get_message_completion(self, model: str, content):
        async for chunk in self.get_streaming_completion(
            model, [{"role": "user", "content": content}]
        ):
            yield chunk

    async def get_completion(self, model: str, messages):
        # Updated for v0.5 - using the new function signature
        response = await generate_chat_completion(
            self.__request__,
            {"model": model, "messages": messages, "stream": False}, 
            self.__user__
        )

        return self.get_response_content(response)

    async def stream_prompt_completion(self, prompt, **format_args):
        complete = ""
        async for chunk in self.get_message_completion(
            self.__model__,
            prompt.format(**format_args),
        ):
            complete += chunk
            await self.emit_message(chunk)
        return complete

    async def generate_thought(self, answer):
        return await self.stream_prompt_completion(
            thoughts_prompt, answer=answer, question=self.__question__
        )

    async def analyze_iteration(self, best_answer, best_score):
        return await self.stream_prompt_completion(
            analyze_prompt,
            question=self.__question__,
            best_answer=best_answer,
            best_score=best_score,
        )

    async def update_approach(self, answer, improvements):
        return await self.stream_prompt_completion(
            update_prompt,
            question=self.__question__,
            answer=answer,
            improvements=improvements,
        )

    async def evaluate_answer(self, answer):
        result = await self.stream_prompt_completion(
            eval_answer_prompt,
            answer=answer,
            question=self.__question__,
        )
        try:
            score = re.search(r"\d+", result).group()
            return int(score)
        except AttributeError:
            logger.error(f'AnswerEval: unable to parse "{result[:100]}"')
            return 0

    def get_response_content(self, response):
        try:
            # Handle different response formats
            if isinstance(response, dict):
                return response["choices"][0]["message"]["content"]
            elif hasattr(response, 'choices'):
                return response.choices[0].message.content
            else:
                return str(response)
        except (KeyError, IndexError, AttributeError):
            logger.error(
                f'ResponseError: unable to extract content from "{str(response)[:100]}"'
            )
            return ""

    def get_chunk_content(self, chunk):
        try:
            # Handle bytes chunks
            if isinstance(chunk, bytes):
                chunk_str = chunk.decode("utf-8")
            else:
                chunk_str = str(chunk)
                
            if chunk_str.startswith("data: "):
                chunk_str = chunk_str[6:]

            chunk_str = chunk_str.strip()

            if chunk_str == "[DONE]" or not chunk_str:
                return

            try:
                chunk_data = json.loads(chunk_str)
                if "choices" in chunk_data and len(chunk_data["choices"]) > 0:
                    delta = chunk_data["choices"][0].get("delta", {})
                    if "content" in delta:
                        yield delta["content"]
            except json.JSONDecodeError:
                logger.error(f'ChunkDecodeError: unable to parse "{chunk_str[:100]}"')
        except Exception as e:
            logger.error(f'ChunkProcessingError: {str(e)}')
            return`
